### PR TITLE
[RFC] Support different types for the index buffer.

### PR DESCRIPTION
**Pull request by bobtwinkles**
Using `u32` all the time is massively wasteful for many workloads, since we rarely have meshes with more than 65535 vertexes (so we can get away with `u16`s).  GL also supports `u8` index buffers, which are handy for very small meshes, like the full-screen triangles often used for lots of post-processing effects, so we add support for those as well since it's easy enough to do. We may also want to add support for `()` as an "index buffer type" for directly rendered meshes, as right now the user needs to arbitrarily pick a type (from `u8`, `u16`, or `u32`) to fill the `Tess` type parameter

### CODE REVIEW

**Comment by phaazon** [comment]
Yeah, I’ve thought about that as well but never did it. Nice improvement, thank you! Man that’s going to be a massive changelog 😆

File: luminance/src/tess.rs [hunk 1]
```diff
@@ -393,24 +399,54 @@ pub enum TessError {
...
+  fn to_glenum() -> GLenum {
+    gl::UNSIGNED_INT
+  }
+
+  fn size() -> usize {
+    size_of::<u32>()
+  }
+}
+
+pub struct Tess<I> {
```
Comment (phaazon) [inline luminance/src/tess.rs]: In the past (and current crate version), `Tess` had vertex type as phantom typing, so that not need is checked when working on vertices later (e.g. getting a slice mapped to memory to work on, etc.).

With deinterleaving and instancing, that tends to be harder to do and would also prevent dynamically set `Tess` (for instance, in a tool, when the vertex type is defined on-the-fly). Hence, I don’t really know why the type is needed here. Can you summarize your thoughts?

File: luminance/src/tess.rs [hunk 2]
```diff
@@ -393,24 +399,54 @@ pub enum TessError {
   Overflow(usize, usize)
 }
 
-pub struct Tess {
+pub unsafe trait VertIndex {
```
Comment (phaazon) [inline luminance/src/tess.rs]: Maybe we can simplify that trait a lot by introducing only an associated value via a constant, like it’s done with `VertexAttribFmt`:

```rust
trait Index {
  const INDEX_FMT: IndexFmt;
}

// …

enum IndexFmt {
  U8,
  U16,
  U32
}
```

Comment (phaazon) [reply luminance/src/tess.rs]: Then, in the implementation, we can simply have a function that maps `IndexFmt` to the required set of OpenGL attributes. That is going to be needed to enable backends (other versions of GL, WebGL, GL ES, maybe Vulkan, etc.)

File: luminance/src/tess.rs (hunk 1 above)
Comment (bobtwinkles) [reply luminance/src/tess.rs]: The idea with using a type parameter here was to avoid a run-time check. If you want to keep `Tess` parameter-free, I can remove it easily.

Comment (phaazon) [reply luminance/src/tess.rs]: So it’s the same initial reason. I think it should be okay to keep it, as there’s no interleaving implied with indices.

Comment (phaazon) [reply luminance/src/tess.rs]: I had made it work with vertices as well a few weeks ago, but the public API is hard to use and won’t work with the correct builder-pattern `TessBuilder`. I think we should be able to merge that quickly.

File: luminance/src/tess.rs (hunk 2 above)
Comment (phaazon) [reply luminance/src/tess.rs]: Also, I think we can name that `TessIndex`, as those are indexes used to build up tessellations.

**Comment by bobtwinkles** [comment]
I've push some commits that remove the type parameter and renamed the trait. I also added support for primitive restart, since I needed it and I was in the area. I can pull that out into a separate PR if requested though.

**Comment by phaazon** [comment]
Neat, thanks! Sorry I wasn’t really around lately, I broke my right wrist while snowboarding… 😕 I’ll have a look at your changes so that we can merge that as soon as possible.

**Comment by phaazon** [comment]
There are some stuff to change, but I’ll do it (I need to code a bit :D) and it’s not really important. Thank you, I’ll merge that today! Great work!

**Comment by bobtwinkles** [comment]
No worries, I also got distracted by some other stuff. Thanks for the review and cleanup!
